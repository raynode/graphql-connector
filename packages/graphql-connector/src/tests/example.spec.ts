import {
  applyFilterParser,
  createBaseSchemaGenerator,
  createModelMapper,
  createSchema,
  FilterParser,
  TypeMapper,
} from '..'

import { setupDatabase } from './my-data'
import { Database, DBTypeRecord, Instance, Model, models as myModels } from './my-models'
import * as myTypes from './my-types'

import { graphql, GraphQLID, GraphQLInt, GraphQLString, printSchema } from 'graphql'
import { DateType } from './graphql-date-type'

const debug = false
// @TODO remember to remove the try-catch

// myModels is expected to be { user: DBUserModel, post: DBPostModel, comment: DBCommentModel }
type Models = typeof myModels

// myTypes is expected to be { DBIntType: any, DBStringType: any, ... }
type Types = myTypes.DBType

// utility to check if some type is a list-type
const isListType = (type: myTypes.DBType): type is myTypes.DBListType => type instanceof myTypes.DBListType

// utility to resolve the final type of a type
const getFinalType = (type: myTypes.DBType) => (isListType(type) ? type.subtype : type)

// utility to create a filter function used in findOne & findMany queries
const buildFilterFn = (data: Record<string, any> = {}) => (instance: Instance<any, any>) =>
  Object.keys(data).reduce(
    (valid, key) =>
      valid && data[key] instanceof Filter
        ? data[key].apply(instance, instance.get(key))
        : instance.get(key) === data[key],
    true,
  )

type FilterFunc = 'eq' | 'has'
interface FilterRule {
  func: FilterFunc
  value: any
  not: boolean
}

class Filter<Key extends keyof Models> {
  public static create<Key extends keyof Models>(old: Filter<Key> | any, model: Models[Key], field: string) {
    if (old instanceof Filter) return old
    return new Filter(model, field)
  }

  private rules: FilterRule[] = []
  private type: myTypes.DBType
  private attribute: Record<string, myTypes.DBType>

  public constructor(private model: Models[Key], private field: string) {
    this.attribute = this.model.attributes[field]
    this.type = getFinalType(this.attribute)
  }
  public add(func: FilterFunc, value: any, not = false) {
    this.rules.push({ func, value, not })
  }
  public apply(instance: Instance<any, any>, compare: any) {
    return this.rules.reduce((valid, { func, not, value }) => {
      if (!valid) return false
      switch (func) {
        case 'has':
          const type = this.type as myTypes.DBIDType
          const model = myModels[type.source]
          const data = model.findMany(buildFilterFn({ [type.sourceAttribute]: instance.get(type.targetAttribute) }))
          return !data.length === not
        case 'eq':
          return (instance.get(this.field) === value) !== not
      }
      // throw new Error(func + ' not defined')
      return valid
    }, true)
  }

  public inspect() {
    return `${this.model.name}(${this.rules.map(rule => rule.func).join(',')})`
  }
}

describe('the example code', async () => {
  try {
    // the filter parser is not strictly necessary, it is available to you to convert the incoming mutation data
    // or the filtered where fields into usable data.
    // there are 3 different modes to this: 'create', 'update' and 'where'
    // these correspond to the mutations create & update and the where filter-input
    const filterParser: FilterParser<Types, Models> = (mode, model, name, value, data) => {
      const attribute = model.attributes[name]
      const type = getFinalType(attribute)
      // in case we try to run a where filter:
      if (mode === 'where') {
        let filter: Filter<any>

        if (name.includes('_')) {
          const [pre, post] = name.split('_')
          const field = pre === 'has' || pre === 'matches' ? post : pre
          const fn: FilterFunc = (pre === 'has' || pre === 'matches' ? pre : post) as any

          filter = Filter.create(data[field], model, field)
          filter.add(fn, value)
        } else {
          filter = Filter.create(data[name], model, name)
          filter.add('eq', value)
        }
        return {
          ...data,
          [name]: filter,
        }
      }
      // in case of an update or create query, we need to convert all associations into the IDs
      if (type instanceof myTypes.DBIDType) {
        const model: Model<any> = myModels[type.source]
        const result = model.findOne(buildFilterFn(value))
        if (!result) throw new Error(`Could not create a new ${model.name} as ${name} could not be set`)

        return {
          ...data,
          [type.targetAttribute]: result.get('id'),
        }
      }
      data[name] = value
      return data
    }

    const findManyPaginated = async (model: Model<any>, filter: ReturnType<typeof buildFilterFn>) => {
      const nodes = await model.findMany(filter)
      return {
        page: null,
        nodes,
      }
    }
    // The modelMapper is responsible to convert a my-model into a graphql-connector model
    // a graphql-connector model needs attributes, associations
    // each association and attribute could be setup with an resolver here
    const modelMapper = createModelMapper<Types, Models>((model, addAttribute, addAssociation) => {
      // to convert a my-model into a graphql-connector model,
      // we need to tell the system all attributes and associations
      Object.keys(model.attributes).forEach(name => {
        // for the attribute
        const attribute = model.attributes[name]
        // get the type of the attribute
        const type = getFinalType(attribute)
        const list = isListType(attribute)
        // check if it is some connection (this system uses IDTypes to connect one model to another)
        if (name !== 'id' && type instanceof myTypes.DBIDType) {
          // create a filter function to be used in the source selectors
          const filterFn = (instance: Instance<any, any>) => {
            const attr = instance.get(type.targetAttribute)
            return (item: Instance<any, any>) => item.get(type.sourceAttribute) === attr
          }
          // create a resolver that handles the model resolve
          const resolver = async (instance: any, args, context) => {
            const source = myModels[type.source]
            if (!list) return source.findOne(filterFn(instance))
            return findManyPaginated(source, filterFn(instance))
          }
          // add this as an accociation, the type.source type conversion is needed as the types do not know this type
          return addAssociation({ list, model: type.source as keyof Models, name, resolver })
        }
        // create a resolver for attributes (default would be instance[name])
        // the example system uses a different style
        const resolver = instance => instance.get(name)
        // add the attribute
        return addAttribute<Instance<any, any>>({ list, name, resolver, type })
      })

      // create the base resolvers
      // only implemented the findOne, findMany and create resolver as I am lazy
      return {
        findMany: async (_, { where, order }) => findManyPaginated(model, buildFilterFn(where)),
        create: async (_, { data }) => (model.create as any)(data),
        findOne: async (_, { where }) => model.findOne(buildFilterFn(where)),
        delete: async () => null,
        update: async () => null,
      }
    })

    // this function tells the generator how to convert my-type into a GraphQLType
    const typeMapper: TypeMapper<Types, Models> = ({ type }) => {
      if (type instanceof myTypes.DBIDType) return GraphQLID
      if (type instanceof myTypes.DBDateType) return DateType
      if (type instanceof myTypes.DBIntType) return GraphQLInt
      if (type instanceof myTypes.DBStringType) return GraphQLString
      throw new Error('invalid type: ' + typeof type)
    }

    // Generate a baseSchemaGenerator by providing the modelMapper and typeMapper
    // here we could change other things, like the nameingStrategy
    const baseSchemaGenerator = createBaseSchemaGenerator({
      modelMapper,
      typeMapper,
      filterParser,
    })

    // apply all my-models to the generator
    // the baseSchema will have all queryFields, mutationsFields and subscriptionFields
    const baseSchema = baseSchemaGenerator(myModels)

    // // this part is graphql only, not more graphql-connector
    const schema = createSchema(baseSchema)

    it('should have the correct schema', () => {
      expect(printSchema(schema)).toMatchSnapshot()
    })

    beforeAll(setupDatabase)

    const runQuery = async (query: string) => graphql(schema, query, null, null)

    it('should find the user by name', async () => {
      const { data, errors } = await runQuery(`{
        authors: Users { nodes {
          name
          email
          posts { nodes {
            title
            text
            upvotes
            comments { nodes {
              msg
              commentor { name }
            } }
          } }
        } }
      }`)
      expect(data.authors).toMatchSnapshot()
    })

    it('should find all posts', async () => {
      const { data } = await runQuery(`{
        Posts { nodes {
          title
          text
          upvotes
          author { name }
          comments { nodes {
            msg
            commentor { name }
          }}
        }}
      }`)

      expect(data).toMatchSnapshot()
    })

    it('should add a new question to the Database', async () => {
      const { data, errors } = await runQuery(`mutation newQuestion {
        createQuestion(data: {
          question: "Do you like this project?"
          answers: ["Yes!", "It looks nice...", "I don't know", "What?", "42!", "Probably not"]
          author: { name: "Georg" }
        }) {
          id
          question
          answers
          author { name email }
        }
      }`)
      expect(data).toMatchSnapshot()
      expect(Database.database.find(entry => entry.get('id') === data.createQuestion.id)).toMatchSnapshot()
    })

    it('should find more data', async () => {
      const { data } = await runQuery(`{
        Users {
          nodes {
            name
            email
            questions { nodes {
              question
              answers
            }}
            posts { nodes {
              title
              upvotes
              comments { nodes {
                msg
                commentor { name }
              }}
            }}
            comments { nodes {
              msg
              post { title }
            }}
          }
        }
      }`)
      expect(data).toMatchSnapshot()
    })

    it('should add Frank as a member', async () => {
      const { data } = await runQuery(`mutation newQuestion {
        frank: createUser(data: {
          name: "Frank"
          group: "member"
          email: "frank@example.com"
        }) {
          id
          name
          email
          group
        }
      }`)
      expect(data).toMatchSnapshot()
      expect(Database.database.find(entry => entry.get('id') === data.frank.id)).toMatchSnapshot()
    })

    it('should find Georg & Paul', async () => {
      const { data } = await runQuery(`{
        georg: User(where: {
          name: "Georg",
        }) {
          name
          email
          questions { nodes {
            question
            answers
          }}
        }
        paul: User(where: {
          name: "Paul",
        }) {
          name
          email
          questions { nodes {
            question
            answers
          }}
        }
      }`)
      expect(data).toMatchSnapshot()
    })

    it('should find all members in name order ', async () => {
      const { data } = await runQuery(`{
        members: Users(where: {
          group: "member"
        }, order: name_ASC) { nodes {
          name
        }}
      }`)
      expect(data).toMatchSnapshot()
    })

    it('should find all members in reversed name order ', async () => {
      const { data } = await runQuery(`{
        members: Users(where: {
          group: "member"
        }, order: name_DESC) { nodes {
          name
        }}
      }`)
      expect(data).toMatchSnapshot()
    })

    it('should find all users with posts', async () => {
      const { data } = await runQuery(`{
        authors: Users(where: {
          has_posts: true
        }) { nodes {
          name
        }}
      }`)
      expect(data).toMatchSnapshot()
    })
  } catch (err) {
    console.error('Error in test!', err)
  }
  it('should run stuff', () => {
    expect(true).toBeTruthy()
  })
})
