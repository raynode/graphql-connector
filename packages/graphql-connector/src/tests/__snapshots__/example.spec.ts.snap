// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the example code should add Frank as a member 1`] = `
Object {
  "frank": Object {
    "email": "frank@example.com",
    "group": "member",
    "id": "User:7",
    "name": "Frank",
  },
}
`;

exports[`the example code should add Frank as a member 2`] = `
Instance {
  "attributes": Object {
    "createdAt": 2018-09-30T22:00:00.000Z,
    "email": "frank@example.com",
    "group": "member",
    "id": "User:7",
    "lastUpdate": 2018-09-30T22:00:00.000Z,
    "name": "Frank",
  },
  "model": Model {
    "attributes": Object {
      "comments": DBListType {
        "subtype": DBIDType {
          "source": "Comment",
          "sourceAttribute": "commentor",
          "target": "User",
          "targetAttribute": "id",
        },
      },
      "createdAt": DBDateType {},
      "email": DBStringType {},
      "group": DBStringType {},
      "id": DBIDType {
        "source": "User",
        "sourceAttribute": "id",
        "target": null,
        "targetAttribute": "id",
      },
      "lastUpdate": DBDateType {},
      "name": DBStringType {},
      "num": DBIntType {},
      "password": DBStringType {},
      "posts": DBListType {
        "subtype": DBIDType {
          "source": "Post",
          "sourceAttribute": "userId",
          "target": "User",
          "targetAttribute": "id",
        },
      },
      "questions": DBListType {
        "subtype": DBIDType {
          "source": "Question",
          "sourceAttribute": "userId",
          "target": "User",
          "targetAttribute": "id",
        },
      },
    },
    "name": "User",
  },
}
`;

exports[`the example code should add a new question to the Database 1`] = `
Object {
  "createQuestion": Object {
    "answers": Array [
      "Yes!",
      "It looks nice...",
      "I don't know",
      "What?",
      "42!",
      "Probably not",
    ],
    "author": Object {
      "email": "georg@example.com",
      "name": "Georg",
    },
    "id": "Question:6",
    "question": "Do you like this project?",
  },
}
`;

exports[`the example code should add a new question to the Database 2`] = `
Instance {
  "attributes": Object {
    "answers": Array [
      "Yes!",
      "It looks nice...",
      "I don't know",
      "What?",
      "42!",
      "Probably not",
    ],
    "createdAt": 2018-09-30T22:00:00.000Z,
    "id": "Question:6",
    "lastUpdate": 2018-09-30T22:00:00.000Z,
    "question": "Do you like this project?",
    "userId": "User:1",
  },
  "model": Model {
    "attributes": Object {
      "answers": DBListType {
        "subtype": DBStringType {},
      },
      "author": DBIDType {
        "source": "User",
        "sourceAttribute": "id",
        "target": "Question",
        "targetAttribute": "userId",
      },
      "createdAt": DBDateType {},
      "id": DBIDType {
        "source": "Question",
        "sourceAttribute": "id",
        "target": null,
        "targetAttribute": "id",
      },
      "lastUpdate": DBDateType {},
      "question": DBStringType {},
      "userId": null,
    },
    "name": "Question",
  },
}
`;

exports[`the example code should find Georg & Paul 1`] = `
Object {
  "georg": Object {
    "email": "georg@example.com",
    "name": "Georg",
    "questions": Object {
      "nodes": Array [
        Object {
          "answers": Array [
            "Yes!",
            "It looks nice...",
            "I don't know",
            "What?",
            "42!",
            "Probably not",
          ],
          "question": "Do you like this project?",
        },
      ],
    },
  },
  "paul": Object {
    "email": "paul@example.com",
    "name": "Paul",
    "questions": Object {
      "nodes": Array [],
    },
  },
}
`;

exports[`the example code should find all members 1`] = `
Object {
  "members": Object {
    "nodes": Array [
      Object {
        "name": "Georg",
      },
      Object {
        "name": "Frank",
      },
    ],
  },
}
`;

exports[`the example code should find all posts 1`] = `
Object {
  "Posts": Object {
    "nodes": Array [
      Object {
        "author": Object {
          "name": "Paul",
        },
        "comments": Object {
          "nodes": Array [],
        },
        "text": "This is a good time to be alive, this is our new Blog. Clean code to the max!",
        "title": "Welcome to the new Blog!",
        "upvotes": 17,
      },
      Object {
        "author": Object {
          "name": "Georg",
        },
        "comments": Object {
          "nodes": Array [
            Object {
              "commentor": Object {
                "name": "Georg",
              },
              "msg": "I believe!",
            },
          ],
        },
        "text": "I think that is a fair question, GraphQL sometimes feels very alien to people.",
        "title": "Was alien technology involved?",
        "upvotes": 4,
      },
    ],
  },
}
`;

exports[`the example code should find all users with posts 1`] = `
Object {
  "authors": Object {
    "nodes": Array [
      Object {
        "name": "Georg",
      },
      Object {
        "name": "Paul",
      },
    ],
  },
}
`;

exports[`the example code should find more data 1`] = `
Object {
  "Users": Object {
    "nodes": Array [
      Object {
        "comments": Object {
          "nodes": Array [
            Object {
              "msg": "I believe!",
              "post": Object {
                "title": "Was alien technology involved?",
              },
            },
          ],
        },
        "email": "georg@example.com",
        "name": "Georg",
        "posts": Object {
          "nodes": Array [
            Object {
              "comments": Object {
                "nodes": Array [
                  Object {
                    "commentor": Object {
                      "name": "Georg",
                    },
                    "msg": "I believe!",
                  },
                ],
              },
              "title": "Was alien technology involved?",
              "upvotes": 4,
            },
          ],
        },
        "questions": Object {
          "nodes": Array [
            Object {
              "answers": Array [
                "Yes!",
                "It looks nice...",
                "I don't know",
                "What?",
                "42!",
                "Probably not",
              ],
              "question": "Do you like this project?",
            },
          ],
        },
      },
      Object {
        "comments": Object {
          "nodes": Array [],
        },
        "email": "paul@example.com",
        "name": "Paul",
        "posts": Object {
          "nodes": Array [
            Object {
              "comments": Object {
                "nodes": Array [],
              },
              "title": "Welcome to the new Blog!",
              "upvotes": 17,
            },
          ],
        },
        "questions": Object {
          "nodes": Array [],
        },
      },
    ],
  },
}
`;

exports[`the example code should find the user by name 1`] = `
Object {
  "nodes": Array [
    Object {
      "email": "georg@example.com",
      "name": "Georg",
      "posts": Object {
        "nodes": Array [
          Object {
            "comments": Object {
              "nodes": Array [
                Object {
                  "commentor": Object {
                    "name": "Georg",
                  },
                  "msg": "I believe!",
                },
              ],
            },
            "text": "I think that is a fair question, GraphQL sometimes feels very alien to people.",
            "title": "Was alien technology involved?",
            "upvotes": 4,
          },
        ],
      },
    },
    Object {
      "email": "paul@example.com",
      "name": "Paul",
      "posts": Object {
        "nodes": Array [
          Object {
            "comments": Object {
              "nodes": Array [],
            },
            "text": "This is a good time to be alive, this is our new Blog. Clean code to the max!",
            "title": "Welcome to the new Blog!",
            "upvotes": 17,
          },
        ],
      },
    },
  ],
}
`;

exports[`the example code should have the correct schema 1`] = `
"type Comment implements Node {
  id: ID
  createdAt: Date
  lastUpdate: Date
  msg: String
  commentor: User
  post: Post
}

input CommentFilter {
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  msg: String
  msg_not: String
  msg_in: [String!]
  msg_not_in: [String!]
}

type Comments implements List {
  nodes: [Comment!]!
  page: Page!
}

enum CommentSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  lastUpdate_ASC
  lastUpdate_DESC
  msg_ASC
  msg_DESC
  commentor_ASC
  commentor_DESC
  post_ASC
  post_DESC
}

input CommentWhere {
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  msg: String
  msg_not: String
  msg_in: [String!]
  msg_not_in: [String!]
  has_commentor: Boolean
  matches_commentor: UserFilter
  has_post: Boolean
  matches_post: PostFilter
}

input CreateCommentData {
  id: ID
  createdAt: Date
  lastUpdate: Date
  msg: String
  commentor: UserWhere
  post: PostWhere
}

input CreatePostData {
  id: ID
  createdAt: Date
  lastUpdate: Date
  title: String
  text: String
  upvotes: Int
  author: UserWhere
  comments: CommentWhere
}

input CreateQuestionData {
  id: ID
  createdAt: Date
  lastUpdate: Date
  answers: [String!]
  question: String
  author: UserWhere
}

input CreateUserData {
  id: ID
  createdAt: Date
  lastUpdate: Date
  name: String
  email: String
  password: String
  group: String
  num: Int
  posts: PostWhere
  comments: CommentWhere
  questions: QuestionWhere
}

\\"\\"\\"
A special custom Scalar type for Dates that converts to a ISO formatted string 
\\"\\"\\"
scalar Date

interface List {
  nodes: [Node!]!
  page: Page!
}

type Mutation {
  createUser(data: CreateUserData): User
  updateUser(data: UpdateUserData, where: UserWhere): User
  deleteUsers(where: UserWhere, order: UserSortOrder): Users!
  createPost(data: CreatePostData): Post
  updatePost(data: UpdatePostData, where: PostWhere): Post
  deletePosts(where: PostWhere, order: PostSortOrder): Posts!
  createComment(data: CreateCommentData): Comment
  updateComment(data: UpdateCommentData, where: CommentWhere): Comment
  deleteComments(where: CommentWhere, order: CommentSortOrder): Comments!
  createQuestion(data: CreateQuestionData): Question
  updateQuestion(data: UpdateQuestionData, where: QuestionWhere): Question
  deleteQuestions(where: QuestionWhere, order: QuestionSortOrder): Questions!
}

interface Node {
  id: ID
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Post implements Node {
  id: ID
  createdAt: Date
  lastUpdate: Date
  title: String
  text: String
  upvotes: Int
  author: User
  comments: Comments
}

input PostFilter {
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  title: String
  title_not: String
  title_in: [String!]
  title_not_in: [String!]
  text: String
  text_not: String
  text_in: [String!]
  text_not_in: [String!]
  upvotes: Int
  upvotes_not: Int
  upvotes_gt: Int
  upvotes_lt: Int
}

type Posts implements List {
  nodes: [Post!]!
  page: Page!
}

enum PostSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  lastUpdate_ASC
  lastUpdate_DESC
  title_ASC
  title_DESC
  text_ASC
  text_DESC
  upvotes_ASC
  upvotes_DESC
  userId_ASC
  userId_DESC
  author_ASC
  author_DESC
  comments_ASC
  comments_DESC
}

input PostWhere {
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  title: String
  title_not: String
  title_in: [String!]
  title_not_in: [String!]
  text: String
  text_not: String
  text_in: [String!]
  text_not_in: [String!]
  upvotes: Int
  upvotes_not: Int
  upvotes_gt: Int
  upvotes_lt: Int
  has_author: Boolean
  matches_author: UserFilter
  has_comments: Boolean
  matches_comments: CommentFilter
}

type Query {
  User(where: UserWhere, order: UserSortOrder): User
  Users(where: UserWhere, order: UserSortOrder): Users!
  Post(where: PostWhere, order: PostSortOrder): Post
  Posts(where: PostWhere, order: PostSortOrder): Posts!
  Comment(where: CommentWhere, order: CommentSortOrder): Comment
  Comments(where: CommentWhere, order: CommentSortOrder): Comments!
  Question(where: QuestionWhere, order: QuestionSortOrder): Question
  Questions(where: QuestionWhere, order: QuestionSortOrder): Questions!
}

type Question implements Node {
  id: ID
  createdAt: Date
  lastUpdate: Date
  answers: [String!]
  question: String
  author: User
}

input QuestionFilter {
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  answers_contains: [String!]
  answers_not_contains: [String!]
  question: String
  question_not: String
  question_in: [String!]
  question_not_in: [String!]
}

type Questions implements List {
  nodes: [Question!]!
  page: Page!
}

enum QuestionSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  lastUpdate_ASC
  lastUpdate_DESC
  answers_ASC
  answers_DESC
  question_ASC
  question_DESC
  userId_ASC
  userId_DESC
  author_ASC
  author_DESC
}

input QuestionWhere {
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  answers_contains: [String!]
  answers_not_contains: [String!]
  question: String
  question_not: String
  question_in: [String!]
  question_not_in: [String!]
  has_author: Boolean
  matches_author: UserFilter
}

input UpdateCommentData {
  id: ID
  createdAt: Date
  lastUpdate: Date
  msg: String
  commentor: CommentWhere
  post: CommentWhere
}

input UpdatePostData {
  id: ID
  createdAt: Date
  lastUpdate: Date
  title: String
  text: String
  upvotes: Int
  author: PostWhere
  comments: PostWhere
}

input UpdateQuestionData {
  id: ID
  createdAt: Date
  lastUpdate: Date
  answers: [String!]
  question: String
  author: QuestionWhere
}

input UpdateUserData {
  id: ID
  createdAt: Date
  lastUpdate: Date
  name: String
  email: String
  password: String
  group: String
  num: Int
  posts: UserWhere
  comments: UserWhere
  questions: UserWhere
}

type User implements Node {
  id: ID
  createdAt: Date
  lastUpdate: Date
  name: String
  email: String
  password: String
  group: String
  num: Int
  posts: Posts
  comments: Comments
  questions: Questions
}

input UserFilter {
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  email: String
  email_not: String
  email_in: [String!]
  email_not_in: [String!]
  password: String
  password_not: String
  password_in: [String!]
  password_not_in: [String!]
  group: String
  group_not: String
  group_in: [String!]
  group_not_in: [String!]
  num: Int
  num_not: Int
  num_gt: Int
  num_lt: Int
}

type Users implements List {
  nodes: [User!]!
  page: Page!
}

enum UserSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  lastUpdate_ASC
  lastUpdate_DESC
  name_ASC
  name_DESC
  email_ASC
  email_DESC
  password_ASC
  password_DESC
  group_ASC
  group_DESC
  num_ASC
  num_DESC
  posts_ASC
  posts_DESC
  comments_ASC
  comments_DESC
  questions_ASC
  questions_DESC
}

input UserWhere {
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  email: String
  email_not: String
  email_in: [String!]
  email_not_in: [String!]
  password: String
  password_not: String
  password_in: [String!]
  password_not_in: [String!]
  group: String
  group_not: String
  group_in: [String!]
  group_not_in: [String!]
  num: Int
  num_not: Int
  num_gt: Int
  num_lt: Int
  has_posts: Boolean
  matches_posts: PostFilter
  has_comments: Boolean
  matches_comments: CommentFilter
  has_questions: Boolean
  matches_questions: QuestionFilter
}
"
`;
